### PR TITLE
Add an option to persist data to DETS file

### PR DESCRIPTION
**Pull request by martosaur**
Hi, this PR is to allow Etso to persist data to disk using DETS.

**WHY**
I run my pet projects on $5 Digital Ocean machine. They are normally fairly simple and used by more or less one person so I don't see any reason to spend money or resources on a full-blown DB. I'd like to use Ecto though. Thus this PR.

**WHAT**
The idea is simple. Maybe too simple. If adapter configured this way:
```
config :my_app, Repo,
  persist_to_dets: true,
  dets_folder: "my/storage/"
```
`TableServer` will load an ETS table with data from a file in the folder. On termination, it will dump data from ETS to the file.

Please let me know what you think

### CODE REVIEW

**Comment by Laymer** [comment]
Hi :)

Not sure if I understood all the details here so really guesstimating after a quick read, but you might be interested in the function below : 

[`dets:is_dets_file/1`](https://erldocs.com/18.0/stdlib/dets.html?i=22&search=det#is_dets_file/1)

As I've seen that you need to validate the folder from which your data is imported in :

[validate_dets_folder/1](https://github.com/evadne/etso/blob/b5638e22b15a07f2442ca9855e560435debb4f1b/lib/etso/adapter/meta.ex#L14)

And perhaps if your data is already loaded in memory, you might benefit from `auto_sync`-ing to DETS, and/or specifying that your table should be kept in a `ram_file` by adjusting the options when [initializing](https://erldocs.com/18.0/stdlib/dets.html?i=34&search=det#open_file/2) : 

```
open_file(Name, Args) -> {ok, Name} | {error, Reason}
Name = tab_name()
Args = [OpenArg]
OpenArg = 
    {access, access()} |
    {auto_save, auto_save()} |
    {estimated_no_objects, integer() >= 0} |
    {file, file:name()} |
    {max_no_slots, no_slots()} |
    {min_no_slots, no_slots()} |
    {keypos, keypos()} |
    {ram_file, boolean()} |
    {repair, boolean() | force} |
    {type, type()} |
    {version, version()}
Reason = term()
``` 

If the data you are using to populate ETS tables is structured in a way that allows you to pass arguments such as count estimates, index key positions, etc. I think that you could take advantage of it to gain some performance regardless of the hosting server. Also if the mirroring between the ETS table and its DETS target is done periodically, it could shorten the delay required to dump everything on disk on termination.

Hope this helps :) 

Best,

Igor

**Comment by martosaur** [comment]
@Laymer 

Hi Igor, thanks for the feedback! Once I get strategic feedback from the repo's owner, I'll try to tune dets usage using your tips.

One question tho. Docs for `ets::to_dets/2` suggest that it empties dets table before dumping data into it. Wouldn't it mean that periodical syncing won't bring any performance boost?

**Comment by Laymer** [comment]
Hi :) 

You're welcome, hope I can help 🙂
As far as I understand, it seems that `ets:to_dets/2` flushes an existing DETS table to clean it up before populating it with the ETS data. 

But that is something I would rather think of as a safety measure right at the end of your program. For example, maybe having a periodically synced mirror from ETS<-->DETS during runtime + at separate dump at the end in another replica of the table would provide some good redundancy for the data? 🤔

This requires advice from someone much more knowledgeable than I am I believe, but I hope it helps somehow ^^

Best,

Igor

**Comment by evadne** [comment]
@Laymer @martosaur Thank you so much for the changes. Apologies for late response.

My suggestion:

1. The Persistence layer should be either something that wraps the adapter like the design of hooks in Guardian, or an additional process in the supervision tree started by the adapter. Ecto Adapters are all given an opportunity to start its own bit of the supervision tree, so we could use that mechanism to hold additional logic. See `Etso.Adapter.Supervisor.init/1`.

2. By using a separate process this also lets you listen for signals and gracefully handle termination such as mentioned in https://elixirforum.com/t/graceful-shutdown-on-sigterm/23780

<details>
<summary>So perhaps we can do something like this (diff attached)</summary>

```
diff --git a/lib/etso.ex b/lib/etso.ex
index c343549..e602211 100644
--- a/lib/etso.ex
+++ b/lib/etso.ex
@@ -4,6 +4,7 @@ defmodule Etso do
   """
 
   @type repo :: Ecto.Repo.t()
+  @type hook :: module()
   @type schema :: module()
   @type table :: :ets.tab()
 end
diff --git a/lib/etso/adapter.ex b/lib/etso/adapter.ex
index 231f442..e883018 100644
--- a/lib/etso/adapter.ex
+++ b/lib/etso/adapter.ex
@@ -25,7 +25,8 @@ defmodule Etso.Adapter do
   @doc false
   def init(config) do
     {:ok, repo} = Keyword.fetch(config, :repo)
-    child_spec = __MODULE__.Supervisor.child_spec(repo)
+    hooks = Keyword.get(config, :hooks, [])
+    child_spec = __MODULE__.Supervisor.child_spec(repo, hooks)
     adapter_meta = %__MODULE__.Meta{repo: repo}
     {:ok, child_spec, adapter_meta}
   end
diff --git a/lib/etso/adapter/supervisor.ex b/lib/etso/adapter/supervisor.ex
index 6e65703..b8f3e8b 100644
--- a/lib/etso/adapter/supervisor.ex
+++ b/lib/etso/adapter/supervisor.ex
@@ -6,22 +6,23 @@ defmodule Etso.Adapter.Supervisor do
   """
 
   use Supervisor
-  @spec start_link(Etso.repo()) :: Supervisor.on_start()
+  @spec start_link(Etso.repo(), [Etso.hook()]) :: Supervisor.on_start()
 
   @doc """
   Starts the Supervisor for the given `repo`.
   """
-  def start_link(repo) do
-    Supervisor.start_link(__MODULE__, repo)
+  def start_link(repo, hooks) do
+    Supervisor.start_link(__MODULE__, repo, hooks)
   end
 
   @impl Supervisor
-  def init(repo) do
+  def init(repo, hooks) do
     children = [
       {Etso.Adapter.TableSupervisor, repo},
       {Etso.Adapter.TableRegistry, repo}
     ]
 
+    children = children ++ Enum.map(hooks, & &1.build_children())
     Supervisor.init(children, strategy: :one_for_one)
   end
 end
```

</details>

Where the hook module conforms to some kind of new Etso.Hook behaviour which has a required callback to return child specs. Then you plug your DETS integration in there (wrap in a GenServer) like this.

```elixir
defmodule Etso.Dets.Hook do
  @behaviour Etso.Hook
  def build_children do
    [Etso.Dets.Server]
  end
end
```

Where the actual server does something like this

```elixir
defmodule Etso.Dets.Server do
  use GenServer

  def init(_) do
    # also set timer to regularly save
    {:ok, nil, {:continue, nil}} 
  end

  def handle_continue(_, _) do
    # load
  end

  def terminate(_, _) do
    # save
  end
end
```

The other upside of this is that we will be able to start bundling a bunch of hooks and the customer (user) can pull in whatever they want but only those that they want.

Please let me know your thoughts.

**Comment by evadne** [comment]
Other thoughts:

- If we want to provide `:dets` integration as a first-class feature then I suppose presence of something like `dets_path` implies the user wants to use `:dets`.

- We also need to consider whether this will make sense on schema change. I don’t know what’s going to happen if, say, you build a struct, persist it with `:dets` and attempt to load it again later, after having changed fields in that struct. Within Etso the fields are enumerated then dumped into n-arity tuples so given

```elixir
defmodule Model do
  use Ecto.Schema
  @primary_key {:id, :id, autogenerate: false}

  schema "test" do
    field :foo, :string
  end
end
```

then `%Model{foo: :bar}` is actually stored in ETS in the form of `{(id), :bar}` and if the fields are re-ordered or if new ones were added, then we may have quite a chaotic situation on our hands (note Apple only got to lightweight migration for Core Data after several iterations and they could stand on the shoulders of the giant which is SQLite; in our case we won’t even have original field names).

(We were able to get away with this because we don’t do hot deploys)

So I guess a proper disk-based persistence solution, even though we may want to advertise it as quick & dirty, will not provide a straight dump of the ETS table via DETS, but instead will need to convert each tuple to a more archive-oriented format that is resilient against casual schema changes, and provide the application developer with appropriate error recovery pathways.

In short, implement proper disk archive format first and define how the application handles schema changes. Implement a checkpointer / saver process, set a timer then on each tick call safe_fixtable, traverse the tables (you may do this concurrently, if you wish) and save with dets. Then implement a loader process which loads from dets and invokes the application on exceptions (errors). Then pack it all up in some kind of hook module and get hooks support into Etso proper. That’d be how I approach it.

There might be an easier way to do it though.

**Comment by evadne** [comment]
Finally, regarding on-disk format, a very cheat-y way would be to add a protocol and throw the problem over the wall to the application developer.

```elixir
defprotocol Etso.Dets.Encoding do
  @spec load(term()) :: {:ok, t} | :error
  @spec dump(t) :: {:ok, term()} | :error
  …
end
```

**Comment by Laymer** [comment]
@evadne Thank you very much for your advice, I think that most of what you have described actually applies to persistence mechanisms overall. From experience using a DETS-based persistence layer at one point for a KV-storage module, I could not agree more on the fact that granular resilience is crucial.

For example, I was able to use Lasp with its DETS-based storage backend on my computer and retrieve data repaired even after crashes of nodes and several failures. Deployed the release on GRiSP embedded systems and found out that the built-in repair mechanism was not sufficient.

If the system was interrupted by a hardware reset while DETS was writing, a fatal exception would occur at boot time for all subsequent restarts of the same deployment (i.e. node was permanently lost and manually re-deploying was required).

I am not really experienced enough to tell, but this actually does sound intuitive for me : 

> Finally, regarding on-disk format, a very cheat-y way would be to add a protocol and throw the problem over the wall to the application developer.

If `etso` could be seen as a  _frame_ for the data to be drawn in, a supervision tree to which user-written modules could be attached make sense imo. @evadne if I understand correctly the design you have proposed, the mental model that I could build is that the persistence layer is the product of a generic resiliency layer and optional user-defined operations. I think this is a nice balance between safety and fault tolerance and customization.

Not sure if all of the above is very clear and understandable, but tl;dr would be : 
I think supervision & custom modules for DETS are a good idea

**Comment by evadne** [comment]
@martosaur yo

**Comment by martosaur** [comment]
@evadne Hey! Sorry for taking so long to get back to this, it's busy times.

Ok, so the feature is certainly more complex than I initially thought. I like the idea of implementing persistence to file as a hook, hopefully, I'll get too it sooner or later. 

As for the proper serialization and DETS support, I doubt I have enough experience to handle this.   I'll close this PR for now until I come up with a better one. Sorry for me being that guy and thanks a lot for the guidance <3
